### PR TITLE
solana-program - altbn128: add g1 & g2 compression

### PR DESCRIPTION
**Pull request by ananas-block**
#### Problem
- Groth16 Proofs take up 256 bytes of instruction data per proof (2 g1 64 bytes each and one g2 element of 128 bytes).
- For PSPs we currently need to verify two Groth16 proofs.
- With g1 and g2 compression we can half the bytes per proof to 128.
- Elliptic curve points consist of X and Y coordinates for every X there are two possible Y coordinates, one positive and one negative. Thus, you can send the X coordinate and recover the corresponding Y coordinates. You can use a bitmask on the X coordinate to convey whether to pick the positive or the negative Y coordinate.
- ark-works 0.4. features functions to compress and decompress g1 and g2 elements as described above

#### Summary of Changes
- added alt bn128 g1 and g2 syscall
- used the existing ark works dependencies introduced with the alt bn syscall in https://github.com/solana-labs/solana/pull/27961

#### Implementation Notes

- [This code](https://github.com/ananas-block/solana/pull/1/files#diff-3617e23d40791b23531dc3bdab64afc5c8b783a712393b777e44f2ff41b9e3f4R1849-R1892) is repetitive to enable different sized output fixed size arrays. I am using fixed sized arrays to avoid vectors since we have been repeatedly running low on heap memory in our programs. 

#### Performance

g1 compress: 1 * 1000 / 33 = 30 CU
g1 decompress: 13.154 * 1000 / 33 = 398 CU
g2 compress: 2.8543 * 1000 / 33 = 86 CU
g2 decompress: 449.13 * 1000 / 33 = 13610 CU

See [this branch](https://github.com/ananas-block/solana/tree/altbn128-compression-bench) for the bench implementation.

g1_decompress           time:   [13.089 us 13.154 us 13.237 us]                           
                        change: [-10.850% -10.279% -9.6768%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 11 outliers among 100 measurements (11.00%)
  1 (1.00%) low mild
  3 (3.00%) high mild
  7 (7.00%) high severe

g1_compress             time:   [999.77 ns 1.0049 us 1.0110 us]                         
                        change: [-3.6976% -3.0299% -2.3047%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 12 outliers among 100 measurements (12.00%)
  1 (1.00%) low mild
  2 (2.00%) high mild
  9 (9.00%) high severe

g2_decompress           time:   [447.25 us 449.13 us 451.51 us]                          
                        change: [-14.401% -13.723% -13.047%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 15 outliers among 100 measurements (15.00%)
  8 (8.00%) high mild
  7 (7.00%) high severe

g2_compress             time:   [2.8443 us 2.8543 us 2.8695 us]                         
                        change: [-6.2482% -5.4817% -4.7391%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 15 outliers among 100 measurements (15.00%)
  5 (5.00%) high mild
  10 (10.00%) high severe

### CODE REVIEW

File: sdk/program/src/alt_bn128_compression.rs [hunk 1]
```diff
@@ -0,0 +1,493 @@
...
+            Validate::No,
+        )
+        .map_err(|_| AltBn128CompressionError::DecompressingG2Failed)?;
+        let mut g2_bytes = [0u8; alt_bn128_compression_size::G2_COMPRESSED];
+        G2::serialize_compressed(&g2, g2_bytes.as_mut_slice())
+            .map_err(|_| AltBn128CompressionError::CompressingG2Failed)?;
+        Ok(convert_endianness::<64, 64>(&g2_bytes))
+    }
+
+    pub fn convert_endianness<const CHUNK_SIZE: usize, const ARRAY_SIZE: usize>(
```
Comment (vadorovsky) [inline sdk/program/src/alt_bn128_compression.rs:170]: I'm wondering if it would be useful to put in in some util module?

Comment (ananas-block) [reply sdk/program/src/alt_bn128_compression.rs:170]: hm I don't see an existing utils module, not sure where to put it

Comment (vadorovsky) [reply sdk/program/src/alt_bn128_compression.rs:170]: https://github.com/solana-labs/solana/blob/master/sdk/src/program_utils.rs seems like a good candidate.

It has only one function, completely unrelated to endianness or byte processing, but... it's used across `programs/**`, `sdk/programs/**`, so I guess it's good for functions shared across SDK.

Comment (ananas-block) [reply sdk/program/src/alt_bn128_compression.rs:170]: it's not in sdk/programs I don't have access to this file

**Pull request closed without merging**